p4-16/psa/PSA.mdk

```diff
@@ -2572,3 +2543,131 @@ Certainly if there are long durations of time when egress decides to
+followed.  We recommend that developers selecting P4 devices ask their
+designers about these issues, if those developers consider the issues
+important for their purposes.
+
+Recommendation 1: Packets that arrive on the same input port should
+begin ingress processing in the same relative order as they arrived.
+
+Recommendation 2: Packets that go out on the same output port should
+begin egress processing in the same relative order as they are
+transmitted.
+
+Recommendation 3: PRE unicast packets (i.e. those that follow the
+"enqueue one packet" path in the pseudocode of section
+[#sec-after-ingress]) that arrived from the same ingress port, did
+ingress processing once (i.e. were not resubmitted or recirculated),
+were sent to the PRE with the same `class_of_service` value, and
+destined for the same egress port, should begin egress processing in
+the same relative order as they began ingress processing.
+
+It is expected that some PSA implementations will implement the class
+of service mechanism by having a separate FIFO queue per class of
+service, and thus while unicast packets with the same ingress port,
+egress port, and class of service will pass through the system in FIFO
+order if they follow all recommendations above, unicast packets with
+the same ingress and egress port, but different classes of service,
+may be processed by the egress control block in a different order than
+they were processed by the ingress control block.
+
+If an implementation satisfies recommendations 1 through 3, then
+unicast traffic assigned to the same class of service will maintain
+its relative order through the device.
+
+Recommendation 4: PRE multicast packets (i.e. those that follow the
+"Make 0 or more copies" path in the pseudocode of section
+[#sec-after-ingress]) that arrived on the same ingress port, did
+ingress processing once, were sent to the PRE with the same pair of
+values (`class_of_service`, `multicast_group`), for copies destined to
+the same egress port, and with the same pair of values for
+(`egress_port`, `instance`), should begin egress processing in the
+same relative order as they began ingress processing.
+
+There is no such expectation for multicast packets with different
+`class_of_service` values, again because of separate queues in the PRE
+for different `class_of_service` values.
+
+It is also understood that for a short period of time after the
+control plane modifies the set of copies to be made for a particular
+`multicast_group` value, that it may be especially difficult to
+satisfy Recommendation 4.  That recommendation is intended to apply
+only when the set of copies to be made for the multicast group has
+remained unchanged for a period of time.
+
+If an implementation satisfies recommendations 1, 2, and 4, then
+multicast traffic assigned to the same class of service will maintain
+its relative order through the device, when multicast group membership
+has been stable for long enough.
+
```

Comment:
I might also mention that no specific assumptions should be made with regards to the relative order between multicast and unicast packets. In most modern PREs they are queued and processed somewhat separate from each other, mostly because it is understood that a multicast packet generally spends more time inside the PRE while it is being replicated N times.

Reply:
I added a few sentences mentioning that there is no such recommendation for relative ordering of unicast vs. multicast packets in PRE, with a couple of reasons to motivate why.